src/libraries/System.Resources.Extensions/src/BinaryFormat/Deserializer/Deserializer.cs

```diff
@@ -0,0 +1,412 @@
+        int pendingCount = _pendingSerializationInfo?.Count ?? 0;
+        while (_pendingSerializationInfo is not null && _pendingSerializationInfo.Count > 0)
+        {
+            PendingSerializationInfo? pending = _pendingSerializationInfo.Dequeue();
+
+            // Using pendingCount to only requeue on the first pass.
+            if (--pendingCount >= 0
+                && _pendingSerializationInfo.Count != 0
+                && _incompleteDependencies is not null
+                && _incompleteDependencies.TryGetValue(pending.ObjectId, out HashSet<int>? dependencies))
+            {
+                // We can get here with nested ISerializable value types.
+
+                // Hopefully another pass will complete this.
+                if (dependencies.Count > 0)
+                {
+                    _pendingSerializationInfo.Enqueue(pending);
+                    continue;
+                }
+
+                Debug.Fail("Completed dependencies should have been removed from the dictionary.");
+            }
+
+            // All _pendingSerializationInfo objects are considered incomplete.
+            pending.Populate(_deserializedObjects, Options.StreamingContext);
+            _pendingSerializationInfoIds?.Remove(pending.ObjectId);
+            ((IDeserializer)this).CompleteObject(pending.ObjectId);
+        }
+
+        if (_incompleteObjects.Count > 0 || (_pendingUpdates is not null && _pendingUpdates.Count > 0))
+        {
+            // This should never happen outside of corrupted data.
+            throw new SerializationException(SR.Serialization_Incomplete);
+        }
+
+        // Notify [OnDeserialized] instance methods for all relevant deserialized objects,
+        // then callback IDeserializationCallback on all objects that implement it.
+        OnDeserialized?.Invoke(Options.StreamingContext);
+        OnDeserialization?.Invoke(null);
+
+        return _deserializedObjects[_rootId];
+    }
+
+    [RequiresUnreferencedCode("Calls DeserializeNew(Id)")]
+    private void DeserializeRoot(Id rootId)
+    {
+        object root = DeserializeNew(rootId);
+        if (root is not ObjectRecordDeserializer parser)
+        {
+            return;
+        }
+
+        _parseStack.Push(rootId);
+        _parserStack.Push(parser);
+
+        while (_parserStack.Count > 0)
+        {
+            ObjectRecordDeserializer? currentParser = _parserStack.Pop();
+            int currentId = _parseStack.Pop();
+            Debug.Assert(currentId == currentParser.ObjectRecord.ObjectId);
+
+            Id requiredId;
+            while (!(requiredId = currentParser.Continue()).IsNull)
+            {
+                // A record is required to complete the current parser. Get it.
+                object requiredObject = DeserializeNew(requiredId);
+                Debug.Assert(requiredObject is not SerializationRecord);
+
+                if (requiredObject is ObjectRecordDeserializer requiredParser)
```

Comment:
An item gets created, asserted it's not one thing, tested to be another, and ignored if it's any "third" thing.

If `requiredObject` is always `ObjectRecordDeserializer` then add an assert, or change to a hard cast.

If it's not, then describe why ignoring anything else is acceptable.

Reply:
Some sort of comment(s) explaining that should be added to the code.  Both why it's asserting `not SerializationRecord` and why "everything else" needs no further handling